### PR TITLE
docs: Check TS in docs using typescript-docs-verifier

### PR DESCRIPTION
**Pull request by maschad**
Closes https://github.com/libp2p/js-libp2p/issues/1228

### CODE REVIEW

**Comment by maschad** [comment]
Thanks for the feedback.

> I think this should be added to aegir as a `doc-check` command though, similar to `dep-check`.
Good point, I'll add it as a feature

> It can manage the tsconfig.json file for the docs then, in the same way we do it for [linting](https://github.com/ipfs/aegir/blob/master/src/lint.js?rgh-link-date=2022-12-06T12%3A30%3A10Z#L65-L88).

A user may want to use a different  `tsconfig.json` in situations [like this](https://github.com/bbc/typescript-docs-verifier/issues/22) for instance, so it may be best to specify that separately.

> It also doesn't currently check any code blocks in the docs here, because they are fenced with `js` as a type and not `typescript` - `ts` doesn't seem to work either, though maybe there's some config for that?

I think it's because of you were traversing the wrong directory in the script [here](https://github.com/libp2p/js-libp2p/pull/1505/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R83) it works otherwise.

I opened a [ticket on aegir](https://github.com/ipfs/aegir/issues/1117) once this is closed we can update and review this.

**Comment by maschad** [comment]
Once https://github.com/ipfs/aegir/pull/1134 is merged and released then this can be reviewed

File: doc/LIMITS.md
Comment (maschad) [inline doc/LIMITS.md:38]: We will have to find a way to either silence the compiler in regards to this  since `number` only refers to a type, but is being used as a value here.

I know that https://github.com/libp2p/js-libp2p/pull/1456 sort to avoid giving these values but I wonder if we should revert to stating the defaults

File: doc/LIMITS.md
Comment (maschad) [inline doc/LIMITS.md:53]: This `export` could be confusing to end-users but I haven't found another way to avoid the error about redeclaring block-scoped variables unless we change the name of the variable, which would be required for each declaration of `node`

File: doc/LIMITS.md
Comment (achingbrain) [reply doc/LIMITS.md:38]: It is being used as a value so the compiler is correct in failing.  To use it as a type the example needs to use an actual number (like `50`) or show the interface definition instead.

I think docs are better if the examples can be copy/pasted

File: doc/LIMITS.md
Comment (achingbrain) [reply doc/LIMITS.md:53]: Usually the `isolatedModules` controls errors in ts code that neither imports nor exports anything - did you try that?

Would rather not have to put `export {}` everywher.

**Review by achingbrain — CHANGES_REQUESTED** [review]
Please can you revert all the formatting changes and stick to the same coding standards as the main codebase.  Single quotes for strings, no semi colons, no extra comma for the final entry in a list, etc.

**Comment by achingbrain** [comment]
The interop suite has a peer dep on aegir@37.x so the dependency tree can't be resolved:

```
npm ERR! Could not resolve dependency:
npm ERR! peer aegir@"^37.2.0" from @libp2p/interop@4.0.2
npm ERR! node_modules/@libp2p/interop
npm ERR!   dev @libp2p/interop@"^4.0.0" from the root project
```

This is blocked until https://github.com/libp2p/interop/pull/82 is merged.

**Comment by achingbrain** [comment]
TBH we may wish to do the aegir upgrade first in it's own PR, then revisit this one once it's merged as adding return types to all the functions will be quite noisy.

File: doc/METRICS.md
Comment (achingbrain) [inline doc/METRICS.md:3]: There is useful information here that isn't in the `@libp2p/prometheus-metrics` docs.

Can you please PR that repo with the information that's been removed here.

You can use the `@packageDocumentation` tag ([example](https://github.com/libp2p/js-libp2p/blob/master/src/pnet/index.ts#L1C1-L23) - [output](https://libp2p.github.io/js-libp2p/modules/pnet.html)) to get general docs to appear on a given page, and the `@example` tag to annotate individual methods, classes or interfaces with code that people can copy/paste to get started.

File: doc/LIMITS.md
Comment (achingbrain) [inline doc/LIMITS.md:272]: This should  follow the same linting rules as the main codebase, so use single quotes.

File: doc/LIMITS.md
Comment (achingbrain) [inline doc/LIMITS.md:262]: The quoting is inconsistent - here it's single quotes, in the `allow` string list below it becomes double quotes.

It should follow the same linting rules as the main codebase, so use single quotes.

File: doc/tsconfig.json
Comment (achingbrain) [inline doc/tsconfig.json:1]: Can we bundle this config with aegir and load the config from there?

It would be better than copy/pasting this file into every repo.

**Review by achingbrain — CHANGES_REQUESTED** [review]

File: doc/METRICS.md
Comment (maschad) [reply doc/METRICS.md:3]: Upon reconsideration I think it makes sense to keep this information here and then also add in the `@packageDocumentation` tags and examples for those who may interact with the metrics lib independently of libp2p core.

File: doc/tsconfig.json
Comment (maschad) [reply doc/tsconfig.json]: Good point, once https://github.com/ipfs/aegir/pull/1255 is merged then we can upgrade `aegir` here to `39.0.6` and then merge.

**Comment by maschad** [comment]
Once https://github.com/ipfs/aegir/pull/1255 is merged then we can update to the newest aegir and then merge this.

**Comment by p-shahi** [comment]
this add deps on prom-metrics in the libp2p repo, will wait for the monorepo work to get in follow up in #824 
have package.json in docs folder to keep it out of libp2p deps

**Comment by p-shahi** [comment]
> this add deps on prom-metrics in the libp2p repo, will wait for the monorepo work to get in follow up in #824 have package.json in docs folder to keep it out of libp2p deps

Unblocked, @maschad adding to this PR itself

File: doc/LIMITS.md
Comment (achingbrain) [inline doc/LIMITS.md:288]: These comments should follow the same linting rules as the main codebase; comma-dangle is disallowed.

```suggestion
    ]
  }
```

File: doc/LIMITS.md
Comment (achingbrain) [inline doc/LIMITS.md:289]: More weird linting going on here, semi colons are disallowed in the main codebase:

```suggestion
})
```

File: packages/libp2p/package.json
Comment (achingbrain) [inline packages/libp2p/package.json:175]: Please remove the extra deps.

File: packages/libp2p/package.json
Comment (achingbrain) [inline packages/libp2p/package.json:180]: Please remove the extra deps.

File: packages/libp2p/package.json
Comment (achingbrain) [inline packages/libp2p/package.json:99]: Please revert.

File: packages/libp2p/package.json
Comment (achingbrain) [inline packages/libp2p/package.json:186]: Please remove the extra deps.

**Review by achingbrain — CHANGES_REQUESTED** [review]
The `/doc` folder should be added to the `workspaces` list in the root `package.json`.

It should have a `package.json` added that contains:

1. The `doc-check` command
2. The deps needed by the doc examples
3. The `dep-check` command to ensure the deps don't get out of date (assuming we can make it run on markdown, might need to be a follow up)

File: doc/LIMITS.md
Comment (maschad) [reply doc/LIMITS.md:288]: So the `doc-check` command which utilizes https://github.com/bbc/typescript-docs-verifier only checks that Typescript code snippets compile, it's not a linter. That could be added in a follow up PR but wasn't the original intention of this.

**Comment by maschad** [comment]
I've added the `doc-check` command to the `docs/` workspace. We will need a follow up PR for: 

- doing a `dep-check` than checks the TS snippets in markdown
- Enforcing linting rules on TS snippets in markdown.

Both of these will be features to be added in https://github.com/ipfs/aegir

File: doc/package.json
Comment (achingbrain) [inline doc/package.json:42]: We don't want to publish this to npm

```suggestion
  },
  "private": true
```

**Pull request merged**
